### ISSUE TITLE
Performance issue - warping/unwarping elements.

### ISSUE THREAD

**Issue by florentbr** [mozilla/geckodriver#729]
I noticed that the driver is struggling to warp and unwarp elements.
It takes around 3 seconds to the driver to return 1371 elements while it takes a few milliseconds in the console.
But the most disturbing is that this time doubles once the page is refreshed.

## Firefox Version
55.0a1 (2017-05-09) (32-bit)

## Platform
Windows 10

## Steps to reproduce -
```python
from selenium import webdriver

driver = webdriver.Firefox()

driver.get('https://www.w3.org/TR/webdriver/#protocol')  # 2994 ms

elems = driver.find_elements_by_css_selector('p')  # 3187 ms  / warps 1371 elements

driver.refresh()                                   # 2101 ms
elems = driver.find_elements_by_css_selector('p')  # 8248 ms  / warps 1371 elements

driver.refresh()                                   # 2252 ms
elems = driver.find_elements_by_css_selector('p')  # 12054 ms / warps 1371 elements

driver.execute_script("return 0;")                 # 16 ms    / unwarps 0 elements
driver.execute_script("return 0;", elems)          # 5207 ms  / unwarps 1371 elements
```

**Comment by andreastt** [comment]
Thanks for reporting this!

Marionette’s known element store code is known to be hairy, primarily because we tend not to invalidate the caches after a remoteness change occurs. When a remoteness change happens, we loose the `<xul:browser>`’s window handle because the `outerWindowID` changes. This subsequently causes us to “forget” the already known elements.

The work on fixing Marionette to deal with remoteness changes is https://bugzilla.mozilla.org/show_bug.cgi?id=1311041, and I suspect it’s best to deal with that before dealing with this because there are a number of interdependencies.

**Comment by whimboo** [comment]
I don't think that this is related to a remoteness change. Keep in mind that the above code simply reloads the same page. So under such a condition we do not see a reload of the listener. 

The time we spent in retrieving the elements increases linear, as it can be seen here:

** iteration 0 with 1371 elements: 1.21816301346s
** iteration 1 with 1371 elements: 3.68121099472s
** iteration 2 with 1371 elements: 6.05406498909s
** iteration 3 with 1371 elements: 8.95627808571s
** iteration 4 with 1371 elements: 12.4877390862s
** iteration 5 with 1371 elements: 14.9554760456s
** iteration 6 with 1371 elements: 18.0221178532s
** iteration 7 with 1371 elements: 21.0832149982s
** iteration 8 with 1371 elements: 23.2673139572s

I filed a bug for Marionette: https://bugzilla.mozilla.org/show_bug.cgi?id=1367325

**Comment by andreastt** [comment]
The implementation of Marionette’s known web element store, found in https://searchfox.org/mozilla-central/rev/2933592c4a01b634ab53315ce2d0e43fccb82181/testing/marionette/element.js#61-194, has pretty obvious performance problems. It was written to be correct, and not to be performant.

We want to investigate why we are not clearing the known element store when the browser navigates to a new document, and also what a better data structure looks like.

We have two scenarios where we need to access the cache:

1. Adding a new element by element object reference
2. Looking up element by UUID

We currently perform a linear search for duplicate elements. A binary search would be much quicker, while a double-linked map would have no performance penalty whatsoever at the expense of doubled memory usage.

We also use an object literal with `XPCNativeWrapper`s. I think we would be able to use a `WeakMap` or a `Map` in this case. Because we can’t _assign_ the UUID web element reference to the `DOMNode`, we need to store that individually (and a `Set` is consequently not an option).

**Comment by florentbr** [comment]
I tested a warper based on `WeakMap` and `Map` with a script injection to see what kind of performance could be expected. The time to get 1371 elements dropped from around **3 seconds** to **70 milliseconds**:

```python
JS_FIND_ALL_BY_CSS = """

return warp(getPageCache(), window.document.querySelectorAll(arguments[0]));

function Cache() {
  this.mapElements = new Map();
  this.mapIds = new WeakMap();
  this.idNext = +((Date.now() + '.0').split('').reverse().join(''));
}

function getPageCache() {

  var cache = window._4f735466 || (window._4f735466 = new Cache());
  
  for (var [id, elem] of cache.mapElements) {
    if (!elem.isConnected) {
      cache.mapElements.remove(id);
      cache.mapIds.remove(elem);
    }
  }
  
  return cache;
}

function warp(cache, elements) {

  var warpedElements = [];
  
  for (var element of elements) {
  
    var id = cache.mapIds.get(element);
    
    if (!id) {
      id = '' + cache.idNext++;  // increment id : {counter}.{reversed timestamp}
      cache.mapIds.set(element, id);
      cache.mapElements.set(id, element);
    }
    
    warpedElements.push({'element-6066-11e4-a52e-4f735466cecf': id});
  }
  
  return JSON.stringify(warpedElements);
}

"""

from selenium import webdriver
import json

driver = webdriver.Firefox()

driver.get('https://www.w3.org/TR/webdriver/#protocol')

elems = driver._unwrap_value(json.loads(driver.execute_script(JS_FIND_ALL_BY_CSS, 'p')))  # 66ms
print(elems[1370].id)  # "1370.6513706275941"

elems = driver._unwrap_value(json.loads(driver.execute_script(JS_FIND_ALL_BY_CSS, 'p')))  # 59ms
print(elems[1370].id)  # "1370.6513706275941"

driver.refresh()

elems = driver._unwrap_value(json.loads(driver.execute_script(JS_FIND_ALL_BY_CSS, 'p')))  # 71ms
print(elems[1370].id)  # "1370.4072116275941"

```

**Comment by andreastt** [comment]
@florentbr That is great work! I didn’t expect you to go ahead and prototype a fix yourself.

If you would like, and you have the time and energy, you’re more than welcome to attempt a rewrite of Marionette’s known web element store. I’d be happy to mentor.

**Comment by florentbr** [comment]
@andreastt, Thanks, I'll give it a try.

From what I've seen from the code in marionette, it's way more complicated than just using a WeakMap/Map since the cache of elements is stored at a session level and because it probably needs to deal with XPC wraped elements.

I've tried some changes by extracting the code from omni.ja and by zipping it again.
Is there a better way to make some experiments with marionette and run some tests?

**Comment by andreastt** [comment]
> I've tried some changes by extracting the code from omni.ja and by zipping it again.
> Is there a better way to make some experiments with marionette and run some tests?

I recommend checking out mozilla-central (either with hg or [git](https://sny.no/2016/03/geckogit)). You can install the necessary dependencies to get started using `./mach bootstrap`. After than you can build with `./mach build` (rebuilds not necessary when working on Marionette).

The Marionette source code lives in testing/marionette, and the file you’re interested in is testing/marionette/element.js. You can use `./mach marionette-test --gecko-log - -vv FILE` to run tests. Examples of tests can be found in testing/marionette/harness/marionette_harness/tests/unit.

I’m known as _ato_ on irc.mozilla.org if you have any questions.

**Comment by andreastt** [comment]
Also sprach Florent Breheret:

> I gave up on setting up the environment to build Firefox. It's way too
> challenging and time consuming.

It should just be a matter of running

	% ./mach bootstrap

on your system of choice (Linux/macOS/Windows).

But thanks for your efforts narrowing down this problem anyway.  It has
been very helpful!  I will take this over and craft a patch that can be
applied directly to mozilla-central.

**Comment by dpalic** [comment]
is there a planned version in which the patch/fix will be integrated?

**Comment by whimboo** [comment]
Looks like this issue was missed. We have had so many changes in the past months that this patch is most likely bitrotted. :/ @andreastt I wonder if we should wait with doing anything here until the window refactoring patch is done and got landed?

**Comment by andreastt** [comment]
I had a look at this patch in the past and I couldn’t make up my mind whether it contained too much complexity or whether it was necessary. The patch is bitrotted at this point but it would be possible to apply with some modifications.

As @whimboo indicated I’m rather busy with a [window tracking refactoring](https://bugzilla.mozilla.org/show_bug.cgi?id=1311041) that has sidetracked me looking into performance patches. I will revisit this once that is done.
